src/de/raw.rs

```diff
@@ -1,475 +1,317 @@
-                        "expected null byte at end of document, got {:#x} instead",
-                        nullbyte
-                    )));
-                }
-            }
-            std::cmp::Ordering::Greater => {
-                return Err(Error::custom(format!(
-                    "document has bytes remaining that were not visited: {}",
-                    length_remaining
-                )));
-            }
-            std::cmp::Ordering::Less => {
-                if length_remaining < 0 {
-                    return Err(Error::custom("length of document was too short"));
-                }
-            }
-        }
-        Ok(())
-    }
-
-    /// Read a string from the BSON.
-    ///
-    /// If utf8_lossy, this will be an owned string if invalid UTF-8 is encountered in the string,
-    /// otherwise it will be borrowed.
-    fn deserialize_str(&mut self) -> Result<Cow<'de, str>> {
-        self.bytes.read_str()
+    pub(crate) fn new(buf: &'de [u8], utf8_lossy: bool) -> Result<Self> {
+        Ok(Self {
+            element: RawElement::toplevel(buf).map_err(Error::deserialization)?,
+            options: DeserializerOptions {
+                utf8_lossy,
+                human_readable: false,
+            },
+        })
     }
 
-    /// Read a null-terminated C style string from the underling BSON.
-    ///
-    /// If utf8_lossy, this will be an owned string if invalid UTF-8 is encountered in the string,
-    /// otherwise it will be borrowed.
-    fn deserialize_cstr(&mut self) -> Result<Cow<'de, str>> {
-        self.bytes.read_cstr()
+    fn value(&self) -> Result<RawBsonRef<'de>> {
+        self.element.value().map_err(Error::deserialization)
     }
 
-    /// Read an ObjectId from the underling BSON.
-    ///
-    /// If hinted to use raw BSON, the bytes of the ObjectId will be visited.
-    /// Otherwise, a map in the shape of the extended JSON format of an ObjectId will be.
-    fn deserialize_objectid<V>(&mut self, visitor: V, hint: DeserializerHint) -> Result<V::Value>
-    where
-        V: serde::de::Visitor<'de>,
-    {
-        let oid = ObjectId::from_reader(&mut self.bytes)?;
-        visitor.visit_map(ObjectIdAccess::new(oid, hint))
-    }
-
-    /// Read a document from the underling BSON, whether it's an array or an actual document.
-    ///
-    /// If hinted to use raw BSON, the bytes themselves will be visited using a special newtype
-    /// name. Otherwise, the key-value pairs will be accessed in order, either as part of a
-    /// [`MapAccess`] for documents or a [`SeqAccess`] for arrays.
-    fn deserialize_document<V>(
-        &mut self,
-        visitor: V,
-        hint: DeserializerHint,
-        document_type: DocumentType,
-    ) -> Result<V::Value>
+    /// Deserialize the element, using the type of the element along with the
+    /// provided hint to determine how to visit the data.
+    fn deserialize_hint<V>(&self, visitor: V, hint: DeserializerHint) -> Result<V::Value>
```

Comment:
Github's diffing, unsurprisingly, got very confused with these changes.  This can be directly compared with the previous version's `deserialize_next` method; in particular, the cases in the the main `match` in both are in the same order so it shouldn't be too hard to scan down them side-by-side.